storage-provider/server/src/p2p/mod.rs

```diff
@@ -1,52 +1,256 @@
-use std::str::FromStr;
+use std::{str::FromStr, sync::Arc, time::Duration};
 
-use libp2p::{identity::Keypair, rendezvous::Namespace, Multiaddr, PeerId};
-use primitives::p2p::keypair_value_parser;
-use register::register;
+use ::blockstore::Blockstore;
+use futures::StreamExt;
+use libp2p::{
+    identify::{self, Event as IdentifyEvent},
+    identity::Keypair,
+    rendezvous::{self, client::Event as RendezvousEvent, Namespace},
+    swarm::{NetworkBehaviour, SwarmEvent},
+    Multiaddr, PeerId, Swarm,
+};
+use primitives::p2p::{keypair_value_parser, DEFAULT_REGISTRATION_TTL};
 use serde::de;
+use swarm::new_swarm;
+use tokio::{select, task::JoinHandle};
 use tokio_util::sync::CancellationToken;
+use tracing::{debug, info, instrument, warn};
 
-mod register;
+pub mod blockstore;
+mod error;
+mod swarm;
 
-pub(crate) use register::RegisterConfig;
+pub use error::P2pError;
 
+/// The time-to-live duration for node registration with rendezvous points.
+const REGISTRATION_TTL: Duration = Duration::from_secs(DEFAULT_REGISTRATION_TTL);
+
+/// Maximum length allowed for a multihash in bytes.
+const MAX_MULTIHASH_LENGTH: usize = 64;
+
+/// Unique namespace used for peer discovery and registration with rendezvous nodes.
 const P2P_NAMESPACE: &str = "polka-storage";
 
-#[derive(Debug, thiserror::Error)]
-pub enum P2PError {
-    #[error(transparent)]
-    Io(#[from] std::io::Error),
-    #[error(transparent)]
-    Dial(#[from] libp2p::swarm::DialError),
-    #[error("Invalid TCP config for swarm")]
-    InvalidTcpConfig,
-    #[error("Invalid behaviour config for swarm")]
-    InvalidBehaviourConfig,
-    #[error(transparent)]
-    TOMLError(#[from] toml::de::Error),
-    #[error("Failed to register at rendezvous point {0}")]
-    RegistrationFailed(PeerId),
-    #[error(transparent)]
-    P2PTransport(#[from] libp2p::TransportError<std::io::Error>),
+/// The protocol version identifier string used by the identify protocol.
+const IDENTIFY_PROTOCOL_VERSION: &str = "polka-storage/1.0.0";
+
+/// Starts a new P2P networking service in a separate tokio task.
+pub fn start_p2p<B>(
+    args: P2pArgs<B>,
+    cancellation_token: CancellationToken,
+) -> Result<JoinHandle<Result<(), P2pError>>, P2pError>
+where
+    B: Blockstore + Send + 'static,
+{
+    // Initialize the p2p worker and move it to the different task
+    let worker = Worker::new(args)?;
+
+    Ok(tokio::spawn(async move {
+        tokio::select! {
+            _ = cancellation_token.cancelled() => {
+                tracing::info!("P2P worker received shutdown signal");
+            }
+            _ = worker.run() => {
+                tracing::info!("P2P worker completed");
+            }
```

Comment:
You're ignoring the returned error by `worker.run()`, at the very least, do something like:
```suggestion
            result = worker.run() => {
            	match result {
            		Ok(_) => tracing::info!("P2P worker completed"),
            		Err(err) => tracing::error!("P2P failed with error: {}", err);
            	}
            }
```